pageserver/src/task_mgr.rs

```diff
@@ -0,0 +1,457 @@
+    task: Arc<PageServerTask>,
+    shutdown_rx: watch::Receiver<bool>,
+    shutdown_process_on_error: bool,
+    future: F,
+) where
+    F: Future<Output = anyhow::Result<()>> + Send + 'static,
+{
+    debug!("Starting task '{}'", task_name);
+
+    let result = SHUTDOWN_RX
+        .scope(
+            shutdown_rx,
+            CURRENT_TASK.scope(task, {
+                // We use AssertUnwindSafe here so that the payload function
+                // doesn't need to be UnwindSafe. We don't do anything after the
+                // unwinding that would expose us to unwind-unsafe behavior.
+                AssertUnwindSafe(future).catch_unwind()
+            }),
+        )
+        .await;
+    task_finish(result, task_name, task_id, shutdown_process_on_error).await;
+}
+
+async fn task_finish(
+    result: std::result::Result<
+        anyhow::Result<()>,
+        std::boxed::Box<dyn std::any::Any + std::marker::Send>,
+    >,
+    task_name: String,
+    task_id: u64,
+    shutdown_process_on_error: bool,
+) {
+    // Remove our entry from the global hashmap.
+    let task = TASKS
+        .lock()
+        .unwrap()
+        .remove(&task_id)
+        .expect("no task in registry");
+
+    let mut shutdown_process = false;
+    {
+        let task_mut = task.mutable.lock().unwrap();
+
+        match result {
+            Ok(Ok(())) => {
+                debug!("Task '{}' exited normally", task_name);
+            }
+            Ok(Err(err)) => {
+                if shutdown_process_on_error {
+                    error!(
+                        "Shutting down: task '{}' tenant_id: {:?}, timeline_id: {:?} exited with error: {:?}",
+                        task_name, task_mut.tenant_id, task_mut.timeline_id, err
+                    );
+                    shutdown_process = true;
+                } else {
+                    error!(
+                        "Task '{}' tenant_id: {:?}, timeline_id: {:?} exited with error: {:?}",
+                        task_name, task_mut.tenant_id, task_mut.timeline_id, err
+                    );
+                }
+            }
+            Err(err) => {
+                if shutdown_process_on_error {
+                    error!(
+                        "Shutting down: task '{}' tenant_id: {:?}, timeline_id: {:?} panicked: {:?}",
+                        task_name, task_mut.tenant_id, task_mut.timeline_id, err
+                    );
+                    shutdown_process = true;
+                } else {
+                    error!(
+                        "Task '{}' tenant_id: {:?}, timeline_id: {:?} panicked: {:?}",
+                        task_name, task_mut.tenant_id, task_mut.timeline_id, err
+                    );
+                }
+            }
+        }
+    }
+
+    if shutdown_process {
+        shutdown_pageserver(1).await;
+    }
+}
+
+// expected to be called from the task of the given id.
+pub fn associate_with(tenant_id: Option<ZTenantId>, timeline_id: Option<ZTimelineId>) {
```

Comment:
While sure it's ok for now, this feels quite odd in the same "single point to orchestrate futures" design, since nothing else is supposed to spawn tasks, that "something" is responsible for associating.

I think we could have changed the code already in this PR to avoid such functions as `pub`, but up to you.

Reply:
You can spawn a task using `task_mgr::spawn` that's initially not associated with any particular tenant or timeline, but the task becomes associated later. That happens in page_service.rs currently. When we accept the connection from the client, it's not associated with any timeline, but when the client sends the `pagestream` command and starts issuing GetPage requests, the task becomes associated with the timeline. We could do that differently, e.g. spawn a new task that is associated with the timeline when the `pagestream` command is received, and let the old task exit. But this seems fine too.

I wouldn't mind refactoring that if you have ideas, in a future PR.

The network protocol with the `pagestream` command feels a bit awkward, we probably should make a bigger refactoring there. Maybe change the protocol from libpq to gRPC, and include the tenant/timeline information earlier, when the connection is authenticated. But that's a different story.